---
layout: vanilla
permalink: /docs/
description: "Documentation for Cell Atlas of Worm."
modified: 2013-09-11
tags: [worm, manual, vignette, atlas]
---
<!DOCTYPE html>
<html lang= "{{ page.lang | default: site.lang | default: "en" }}">

  {% include head.html %}

  <!-- See here for how to use the TOC control:
  https://afeld.github.io/bootstrap-toc/ -->
  <body  data-spy= "scroll" data-target= "#toc">
    {% include header.html %}

    <div class= "container">
      <div class= "row">
        <!-- sidebar, which will move to the top on a small screen -->
        <div class= "col-sm-4">
          <nav id= "toc" data-spy= "affix" data-toggle= "toc"></nav>
        </div>
        <!-- main content area -->
        <div class= "col-sm-8">


<h2>Abstract</h2>
<p> Conventional methods for profiling the molecular content of biological samples fail to resolve heterogeneity that is present at the level of single cells.
  In the past few years, single cell RNA sequencing has emerged as a powerful strategy for overcoming this challenge.
  However, its adoption has been limited by a paucity of methods that are at once simple to implement and cost effective to scale massively.
  Here, we describe a combinatorial indexing strategy to profile the transcriptomes of large numbers of single cells or single nuclei without requiring the physical isolation of each cell (Single cell Combinatorial Indexing RNA-seq or sci-RNA-seq).
  We show that sci-RNA-seq can be used to efficiently profile the transcriptomes of tens-of-thousands of single cells per experiment, and demonstrate that we can stratify cell types from these data.
  Key advantages of sci-RNA-seq over contemporary alternatives such as droplet-based single cell RNA-seq include sublinear cost scaling, a reliance on widely available reagents and equipment, the ability to concurrently process many samples within a single workflow, compatibility with methanol fixation of cells, cell capture based on DNA content rather than cell size, and the flexibility to profile either cells or nuclei.
  As a demonstration of sci-RNA-seq, we profile the transcriptomes of 42,035 single cells from C. elegans at the L2 stage, effectively 50-fold "shotgun cellular coverage" of the somatic cell composition of this organism at this stage.
  We identify 27 distinct cell types, including rare cell types such as the two distal tip cells of the developing gonad, estimate consensus expression profiles and define cell-type specific and selective genes.
  Given that C. elegans is the only organism with a fully mapped cellular lineage, these data represent a rich resource for future methods aimed at defining cell types and states.
  They will advance our understanding of developmental biology, and constitute a major leap towards a comprehensive, single-cell molecular atlas of a whole animal.</p>

<h2>Introduction</h2>

<p>This notebook will show you how to work with the C. elegans L2-stage sci-RNA-seq data from Cao et al., 2017.
It aims to cover the following use cases:</p>

<ol>
  <li> Accessing the raw data
  <li> Exploring the expression pattern of a gene interest
  <li> Finding differentially expressed genes between subsets of cells
  <li> Re-clustering subsets of the data using t-SNE
</ol>

<h2>Installation</h2>

<h3> Required R Packages </h3>
<ul>
  <li> dplyr
  <li> ggplot2
  <li> monocle
</ul>

{% highlight R %}
  suppressPackageStartupMessages({
    library(monocle)
    library(dplyr)
    library(ggplot2)
  })
{% endhighlight %}

<p> This RData file contains both the data and some utility functions to help navigate it. </p>

{% highlight R %}
download.file(
  "http://jpacker-data.s3.amazonaws.com/public/Cao_et_al_2017_vignette.RData",
  destfile = "Cao_et_al_2017_vignette.RData")
{% endhighlight %}

{% highlight R %}
  load("Cao_et_al_2017_vignette.RData")
{% endhighlight %}


<h3> Use of Monocle </h3>

<p> You'll need the R packages listed below. </p>

<p> This vignette was written using Monocle version 2.3.5, which is the version that was used for the paper.
  The source code of Monocle 2.3.5 is provided as a supplementary data file for the paper.
  You can install it using the command: </p>

  {% highlight R %}
    R CMD INSTALL monocle_2.3.5.tar.gz
  {% endhighlight %}

  <p> Later versions of Monocle may produce different results for use case 4 (re-clustering with t-SNE) due to changes in how we preprocess the data before running t-SNE.
    We will update this vignette when we release the next version of Monocle, 2.6.0, to support the new version. </p>

<h3>Important Information</h3>
  <ul>
    <li> cds is a Monocle CellDataSet object containing the single cell RNA-seq data from the main L2-stage C. elegans experiment described in Cao et al., along with annotations.
    <li> cds.neurons is a re-clustered subset of the neuronal cells from cds.
    <li> cds.experiment.2 has data from the second C. elegans experiment described in Cao et al..
      This includes intestine cells that were missed in the first experiment, but the data overall is lower quality than the first experiment.
      In the manuscript, we only included the intestine cells from this experiment and excluded the rest. </li>
    </ul>

<h2>Use Case 1: Accessing the Raw Data</h2>

<p> The raw gene-by-cell UMI (unique molecular identifier) count matrix for a CellDataSet can be accessed using the exprs function. It is stored as a sparse matrix object ("." = 0)
</p>
<p> Note that if you need "even rawer" data (FASTQ files), they are available at the Gene Expression Omnibus (www.ncbi.nlm.nih.gov/geo) under accession code GSE98561.
</p>
<p> If you aren't familiar with working with single cell RNA-seq data, we highly recommend that you take a look at the examples and utility functions presented in the other sections of this document instead of trying to dive in to the raw data directly.
</p>

{% highlight R %}
  exprs(cds)[1:3, 1:3]
{% endhighlight %}

3 x 3 sparse Matrix of class "dgCMatrix" <br>

<table style="overflow:auto;">
    <tr>
      <th> </th>
      <th> cele-001-001.CATGACTCAA </th>
      <th> cele-001-001.AAGACGGCCA </th>
    </tr>
    <tr>
      <th> WBGene00000001 </th>
      <th> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;. </th>
      <th> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;. </th>
    </tr>
    <tr>
      <th> WBGene00000002 </th>
      <th> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;. </th>
      <th> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;. </th>
    </tr>
    <tr>
      <th> WBGene00000003 </th>
      <th> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;. </th>
      <th> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;. </th>
    </tr>
    <tr>
      <th> </th>
      <th> cele-001-001.GCCAACGCCA </th>
      <th> </th>
  </tr>
  <tr>
    <th> WBGene00000001 </th>
    <th> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;. </th>
    <th> </th>
  </tr>
  <tr>
    <th> WBGene00000002 </th>
    <th> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;. </th>
    <th> </th>
  </tr>
  <tr>
      <th> WBGene00000003 </th>
      <th> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;. </th>
      <th> </th>
    </tr>
  </table>

{% highlight R %}
fData(cds)[1:3,]
{% endhighlight %}

<table style="overflow:auto;">
  <tr>
    <th></th>
    <th>gene_id</th>
    <th>symbol</th>
    <th>num_cells_expressed</th>
  </tr>
  <tr>
    <th>WBGene00000001</th>
    <th>WBGene00000001</th>
    <th>aap-1</th>
    <th>1016</th>
  </tr>
  <tr>
    <th> WBGene00000002 </th>
    <th> WBGene00000002 </th>
    <th> aat-1 </th>
    <th> 354 </th>
  </tr>
  <tr>
    <th> WBGene00000003 </th>
    <th> WBGene00000003 </th>
    <th> aat-2 </th>
    <th> 897 </th>
  </tr>
</table>

<p> The <em>pData</em> function is used to access cell annotations.

<ul>
  <li> n.umi is the number of unique molecular identifiers observed to be expressed by a given cell. </li>
  <li> Size_Factor is n.umi divided by the geometric mean of n.umi across all cells. </li>
  <li> tsne_1 and tsne_2 are the coordinates for the cell in the first two t-SNE reduced dimensions. </li>
  <li> Cluster is the cluster id assigned by density peak clustering algorithm applied on the first two t-SNE reduced dimensions. </li>
  <li> cell.type and tissue are the annotations that we used throughout the analysis of Cao et al. </li>
</ul> </p>

{% highlight R %}
pData(cds)[1:3,]
{% endhighlight %}

<style>
  th, td {
    border: 1px;
    padding: 10px;
    text-align: left;
  }
</style>

<table class = "table" style="overflow:auto;">
  <thead>
  <tr>
    <th></th>
    <th>cele-001-001.CATGACTCAA</th>
    <th>cele-001-001.AAGACGGCCA</th>
    <th>cele-001-001.GCCAACGCCA</th>
  </tr>
</thead>
<tbody>
  <tr>
    <th>cell</th>
    <th>cele-001-001.CATGACTCAA</th>
    <th>cele-001-001.AAGACGGCCA</th>
    <th>cele-001-001.GCCAACGCCA	</th>
  </tr>
  <tr>
    <th>n.umi</th>
    <th>144</th>
    <th>790</th>
    <th>832</th>
  </tr>
  <tr>
    <th>plate</th>
    <th>001</th>
    <th>001</th>
    <th>001</th>
  </tr>
  <tr>
    <th>Size_Factor</th>
    <th>0.2368328</th>
    <th>1.2992911</th>
    <th>1.3683674</th>
  </tr>
  <tr>
    <th>num_genes_expressed</th>
    <th>89</th>
    <th>419</th>
    <th>338</th>
  </tr>
  <tr>
    <th>tsne_1</th>
    <th>5.4866377</th>
    <th>-3.8619751</th>
    <th>-0.5594413</th>
  </tr>
  <tr>
    <th>tsne_2</th>
    <th>14.67085</th>
    <th>-27.63448</th>
    <th>41.98569</th>
  </tr>
  <tr>
    <th>Cluster</th>
    <th>20</th>
    <th>6</th>
    <th>13</th>
  </tr>
  <tr>
    <th>peaks</th>
    <th>FALSE</th>
    <th>FALSE</th>
    <th>FALSE</th>
  </tr>
  <tr>
    <th>halo</th>
    <th>TRUE</th>
    <th>TRUE</th>
    <th>TRUE</th>
  </tr>
  <tr>
    <th>delta</th>
    <th>0.02491657</th>
    <th>0.40961274</th>
    <th>0.04445184</th>
  </tr>
  <tr>
    <th>rho</th>
    <th>893.9855</th>
    <th>812.2076</th>
    <th>240.2908</th>
  </tr>
  <tr>
    <th>cell.type</th>
    <th>Unclassified neurons</th>
    <th>Germline</th>
    <th>Intestinal/rectal muscle</th>
  </tr>
  <tr>
    <th>tissue</th>
    <th>Neurons</th>
    <th>Gonad</th>
    <th>Intestinal/rectal muscle</th>
  </tr>
</tbody>
</table>

<br>

<p>neuron.type in pData(cds.neurons) is the annotation used in Figure 4 of Cao et al.</p>

<br>

<table class = "table" style="overflow:auto;">
  <thead>
  <tr>
    <th></th>
    <th>cele-001-001.CATGACTCAA</th>
    <th>cele-001-001.AACTACGGCT</th>
    <th>cele-001-001.GAGGCTTATT</th>
  </tr>
</thead>
<tbody>
  <tr>
    <th>cell</th>
    <th>cele-001-001.CATGACTCAA	</th>
    <th>cele-001-001.AACTACGGCT</th>
    <th>cele-001-001.GAGGCTTATT</th>
  </tr>
  <tr>
    <th>n.umi</th>
    <th>144</th>
    <th>201</th>
    <th>117</th>
  </tr>
  <tr>
    <th>plate</th>
    <th>001</th>
    <th>001</th>
    <th>001</th>
  </tr>
  <tr>
    <th>Size_Factor</th>
    <th>0.2368328</th>
    <th>0.3305791</th>
    <th>0.1924267</th>
  </tr>
  <tr>
    <th>num_genes_expressed</th>
    <th>89</th>
    <th>129</th>
    <th>76</th>
  </tr>
  <tr>
    <th>tsne_1</th>
    <th>0.9574604</th>
    <th>-3.0567593</th>
    <th>-18.5689290</th>
  </tr>
  <tr>
    <th>tsne_2</th>
    <th>0.8288424</th>
    <th>-41.4083795</th>
    <th>-33.9833909</th>
  </tr>
  <tr>
    <th>Cluster</th>
    <th>11</th>
    <th>8</th>
    <th>39</th>
  </tr>
  <tr>
    <th>peaks</th>
    <th>FALSE</th>
    <th>FALSE</th>
    <th>FALSE</th>
  </tr>
  <tr>
    <th>halo</th>
    <th>TRUE</th>
    <th>TRUE</th>
    <th>TRUE</th>
  </tr>
  <tr>
    <th>delta</th>
    <th>0.37046400</th>
    <th>0.25861943</th>
    <th>0.02962754</th>
  </tr>
  <tr>
    <th>rho</th>
    <th>108.71265</th>
    <th>70.88069</th>
    <th>37.29414</th>
  </tr>
  <tr>
    <th>cell.type</th>
    <th>Unclassified neurons</th>
    <th>Ciliated sensory neurons</th>
    <th>Ciliated sensory neurons</th>
  </tr>
  <tr>
    <th>tissue</th>
    <th>Neurons</th>
    <th>Neurons</th>
    <th>Neurons</th>
  </tr>
  <tr>
    <th>neuron.type</th>
    <th>Cholinergic (11)</th>
    <th>ASI/ASJ</th>
    <th>AFD</th>
  </tr>
</tbody>
</table>

<br>

<h2>Use Case 2: Expression Pattern of a Gene of Interest</h2>

<p>The <code>show.expr.info</code> function returns statistics related to the expression of a given gene in tabular form.
  The first argument is the gene name.
  The second argument specifies whether to show statistics at the level of tissues, cell types, or neuron (sub)-types.
  See the examples below. </p>

  <p> the function returns a data frame with five columns: </p>

  <ul>
    <li> facet -- the tissue / cell type / neuron type
    <li> tpm -- the expression of the given gene in the facet in TPM (transcripts per million).
    <li> prop.cells.expr -- the proportion of cells in the facet that express at least one UMI (unique molecular identifier) for the given gene.
      Note that cells in different facets can have very different average number of UMIs per cell, so TPM is the better measure of relative expression.
    <li> n.umi -- the number of UMIs (unique molecular identifiers) observed for the given gene in the facet.
      This is the "sample size" from which the TPM value is computed.
    <li> total.n.umi.for.facet -- the total number of UMIs observed across all genes for all cells in the facet.
  </ul>

  <p> Note that low levels of expression of a given gene in cell types you would not expect may be due to leakage of RNA in the fixed cells.
    For this experiment, based on the expression patterns for a few marker genes, we estimate that the background level of expression you will observe randomly in cells that do not truly express a given gene is ~1-2% of the expression of the highest-expressing cell type. </p>


  {%highlight R%}
  show.expr.info("emb-9", "tissue")
  {%endhighlight%}

  <table class = "table" style="overflow:auto;">
    <thead>
    <tr>
      <th>facet</th>
      <th>tpm</th>
      <th>props.cells.expr</th>
      <th>n.umi</th>
      <th>total.n.umi.for.facet</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Body wall muscle</th>
      <th>7972.27222</th>
      <th>0.96451347</th>
      <th>157028</th>
      <th>19390434</th>
    </tr>
    <tr>
      <th>Gonad</th>
      <th>390.16652</th>
      <th>0.03757116</th>
      <th>3597</th>
      <th>11166871</th>
    </tr>
    <tr>
      <th>Intestine</th>
      <th>597.88569</th>
      <th>0.09775967</th>
      <th>102</th>
      <th>1230975</th>
    </tr>
    <tr>
      <th>Neurons</th>
      <th>89.83882</th>
      <th>0.01406926</th>
      <th>187</th>
      <th>2203067</th>
    </tr>
    <tr>
      <th>Gila</th>
      <th>72.96189</th>
      <th>0.02148228</th>
      <th>39</th>
      <th>787560</th>
    </tr>
    <tr>
      <th>Pharynx</th>
      <th>165.1210</th>
      <th>0.01592357</th>
      <th>14</th>
      <th>85381</th>
    </tr>
    <tr>
      <th>Hypodermis</th>
      <th>47.05950</th>
      <th>0.01821904/th>
      <th>158</th>
      <th>5821384</th>
    </tr>
  </tbody>
  </table>

  {%highlight R%}
    show.expr.info("emb-9", "cell type") %>% head(10)
  {% endhighlight %}

  <table class = "table" style="overflow:auto;">
    <thead>
    <tr>
      <th>facet</th>
      <th>tpm</th>
      <th>prop.cells.expr</th>
      <th>n.umi</th>
      <th>total.n.umi.for.facet</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Distal Tip Cells</th>
      <th>16165.1598</th>
      <th>0.97520661</th>
      <th>3405</th>
      <th>202581</th>
    </tr>
    <tr>
      <th>Body wall muscle</th>
      <th>7972.2722</th>
      <th>0.96451347</th>
      <th>157028</th>
      <th>19390434</th>
    </tr>
    <tr>
      <th>Intestinal/rectal muscle</th>
      <th>5211.8861</th>
      <th>0.84740260</th>
      <th>2622</th>
      <th>439170</th>
    </tr>
    <tr>
      <th>Sex myoblasts</th>
      <th>218.2425</th>
      <th>0.14776632</th>
      <th>75</th>
      <th>377288</th>
    </tr>
    <tr>
      <th>Pharyngeal neurons</th>
      <th>165.1210</th>
      <th>0.01592357</th>
      <th>14</th>
      <th>85381</th>
    </tr>
    <tr>
      <th>Other interneurons</th>
      <th>137.8986</th>
      <th>0.02483070</th>
      <th>17</th>
      <th>172852</th>
    </tr>
    <tr>
      <th>Socket Cells</th>
      <th>123.7517</th>
      <th>0.02793296</th>
      <th>26</th>
      <th>184774</th>
    </tr>
    <tr>
      <th>Coelomocytes</th>
      <th>115.4683</th>
      <th>0.02503682</th>
      <th>71</th>
      <th>544263</th>
    </tr>
    <tr>
      <th>Non-seam hypodermis</th>
      <th>102.4186</th>
      <th>0.02006689</th>
      <th>56</th>
      <th>1059546</th>
    </tr>
    <tr>
      <th>Somatic gonad precursors</th>
      <th>102.1766</th>
      <th>0.06376812</th>
      <th>75</th>
      <th>823856</th>
    </tr>
  </tbody>
  </table>

{%highlight R%}
show.expr.info("R102.2", "neuron type") %>% head(10)
{%endhighlight%}


<table class = "table" style="overflow:auto;">
  <thead>
  <tr>
    <th>facet</th>
    <th>tpm</th>
    <th>prop.cells.expr</th>
    <th>n.umi</th>
    <th>total.n.umi.for.facet</th>
  </tr>
</thead>
<tbody>
  <tr>
    <th>Cluster 21</th>
    <th>9807.74014</th>
    <th>0.91588785</th>
    <th>446</th>
    <th>49587</th>
  </tr>
  <tr>
    <th>Cluster 16</th>
    <th>8266.64832</th>
    <th>0.66025641</th>
    <th>363</th>
    <th>47719</th>
  </tr>
  <tr>
    <th>ASK</th>
    <th>6720.61429</th>
    <th>0.81944444</th>
    <th>155</th>
    <th>24157</th>
  </tr>
  <tr>
    <th>ASI/ASJ</th>
    <th>6482.22237</th>
    <th>0.74358974</th>
    <th>239</th>
    <th>40396</th>
  </tr>
  <tr>
    <th>ASG</th>
    <th>2121.04289</th>
    <th>0.39534884</th>
    <th>48</th>
    <th>26295</th>
  </tr>
  <tr>
    <th>ASEL</th>
    <th>1670.50501</th>
    <th>0.37837838</th>
    <th>17</th>
    <th>11042</th>
  </tr>
  <tr>
    <th>ASER</th>
    <th>795.91129</th>
    <th>0.28571429</th>
    <th>16</th>
    <th>15324</th>
  </tr>
  <tr>
    <th>AWB/AWC</th>
    <th>85.09161</th>
    <th>0.02380952</th>
    <th>4</th>
    <th>23186</th>
  </tr>
  <tr>
    <th>Cholinergic (15)</th>
    <th>60.56935</th>
    <th>0.01538462</th>
    <th>1</th>
    <th>14463</th>
  </tr>
  <tr>
    <th>Pharyngeal (33)</th>
    <th>58.35581</th>
    <th>0.02857143</th>
    <th>2</th>
    <th>25101</th>
  </tr>
</tbody>
</table>

Note that you can see the source code for any of the utility functions used in this vignette.
Just run the function with name no parentheses.

{%highlight R%}
show.expr.info
{%endhighlight%}

{% highlight R %}
  function (gene, expr.info)
  {
      if (class(expr.info) == "character") {
          expr.info = gsub("[.]", " ", tolower(expr.info))
          if (expr.info == "tissue")
              expr.info = tissue.expr.info
          else if (expr.info == "cell type")
              expr.info = cell.type.expr.info
          else if (expr.info == "neuron type")
              expr.info = neuron.type.expr.info
      }
      gene.id =
        get.gene.id(gene, fData.df = expr.info$gene.annotations)
      data.frame(facet = names(expr.info$tpm[gene.id, ]),
      tpm = expr.info$tpm[gene.id,],
      prop.cells.expr = expr.info$prop.cells.expr[gene.id,],
      n.umi = expr.info$n.umi[gene.id, ],
      total.n.umi.for.facet = expr.info$total.n.umi.for.facet)
        %>% arrange(-tpm)
  }
{% endhighlight %}

The <em>plot.expr</em> function can be used to highlight cells that express a given gene on the t-SNE map.
<ul>
  <li> Cells that do not express the gene will be colored grey. They are made semi-transparent so as to better highlight the cells that do express the gene.
  <li> Cells that express the gene will be colored according to their cell type.
    The top 4 highest-expressing cell types will be assigned distinct colors.
    Other cell types will be lumped together.
  <li> Cells in the "Failed QC" category are those that express the gene, but are excluded from the analysis due to either having an unusually low UMI count or being a likely doublet.
</ul>

{% highlight R %}
  plot.expr(cds, "lin-12")
{% endhighlight %}
<div class= "text-center">
  <img src= "{{site.baseurl}}/images/Cao_2017_1.png" width = 600>
</div>

{% highlight R %}
  plot.expr(cds.neurons, "che-3")
{% endhighlight %}

<div class= "text-center">
  <img src= "{{site.baseurl}}/images/Cao_2017_2.png" width = 600>
</div>


<h2>Use Case 3: Differential expression between cell subsets</h2>


<p> We've defined a function <em>two.set.differential.gene.test</em> that finds and reports statistics on differentially expressed genes (DEG) that distinguish between two defined sets of cells.
  This is a wrapper that just adds a bit of functionality around Monocle's <em>differentialGeneTest</em> function. </p>

<em>two.set.differential.gene.test</em> takes four arguments:

<ol>
  <li> cds -- a CellDataSet object that includes both sets of cells you wish to compare
  <li> set.1.filter -- a boolean vector of length ncol(cds) indicating which cells should be in Set 1
  <li> set.2.filter -- a boolean vector of length ncol(cds) indicating which cells should be in Set 2
  <li> formal -- if TRUE, p-values and q-values are computed for differential gene expression and genes are ranked by q-value.
    if FALSE, no formal statistical test is performed and genes are heuristically ranked.
    formal = F makes the function run much, much faster.
</ol>

<b>Warning:</b> if you run <em>two.set.differential.gene.test </em> on large sets of cells (> 1000-ish), it may take a bunch of memory.

<p> The utility functions <em>is.tissue</em>, <em>is.cell.type</em>, and <em>is.neuron.type</em> may be used to create the boolean vectors required for the <em>set.1.filter</em> and <em>set.2.filter</em> parameters.
  Each function takes a CellDataSet and a string and tests for each cell whether its tissue / cell type / neuron type is defined (not NA) and equal to the given value. </p>

{% highlight R %}
is.tissue
{% endhighlight %}

<br>

{% highlight R %}
function (cds, x)
{
    with(pData(cds), !is.na(tissue) & tissue == x)
}
{% endhighlight %}

{% highlight R %}
  head(is.tissue(cds, "Gonad"))
{% endhighlight %}

FALSE TRUE FALSE FALSE FALSE TRUE

{% highlight R %}
  sum(is.tissue(cds, "Gonad"))
{% endhighlight %}
  5628

{% highlight R %}
sum(is.cell.type(cds, "Distal tip cells"))
sum(is.neuron.type(cds.neurons, "ASEL"))
{% endhighlight %}
129
<br>
37

{% highlight R %}
  tissues
{% endhighlight %}
  'Body wall muscle' 'Pharynx' 'Hypodermis' 'Neurons' 'Gila' 'Gonad' 'Intestine'

{% highlight R %}
  cell.types
{% endhighlight %}

'Am/PH sheath cells' 'Body wall muscle' 'Canal associated neurons' 'Cholinergic neurons' 'Ciliated sensory neurons' 'Coelomocytes' 'Distal tip cells' 'Dopaminergic neurons' 'Excretory cells' 'flp-1(+) interneurons' 'GABAergic neurons' 'Germline' 'Intestinal/rectal muscle' 'Intestine' 'Non-seam hypodermis' 'Other interneurons' 'Oxygen sensory neurons' 'Pharyngeal epithelia' 'Pharyngeal gland' 'Pharyngeal muscle' 'Pharyngeal neurons' 'Rectum' 'Seam cells' 'Sex myoblasts' 'Socket cells' 'Somatic gonad precursors' 'Touch receptor neurons' 'Vulval precursors'

{% highlight R %}
  neuron.types
{% endhighlight %}
'AFD' 'ASEL' 'ASER' 'ASG' 'ASI/ASJ' 'ASK' 'AWA' 'AWB/AWC' 'BAG' 'CAN' 'Cholinergic (11)' 'Cholinergic (15)' 'Cholinergic (23)' 'Cholinergic (24)' 'Cholinergic (26)' 'Cholinergic (29)' 'Cholinergic (3)' 'Cholinergic (35)' 'Cholinergic (36)' 'Cluster 10' 'Cluster 13' 'Cluster 16' 'Cluster 17' 'Cluster 21' 'Cluster 25' 'Cluster 27' 'Cluster 40' 'Cluster 5' 'Dopaminergic' 'DVA' 'flp-1(+)' 'GABAergic' 'Pharyngeal (33)' 'Pharyngeal (37)' 'PVC/PVD' 'RIA' 'RIC' 'SDQ/ALN/PLN' 'Touch receptor' 'URX/AQR/PQR'

{% highlight R %}
ASEL.vs.ASER.DEG = two.set.differential.gene.test(
  cds.neurons, is.neuron.type(cds.neurons, "ASEL"),
    is.neuron.type(cds.neurons, "ASER"))
{% endhighlight %}

<em>two.set.differential.gene.test</em> returns the following statistics:
<ul>
  <li> set.1.umi -- the number of UMI (unique molecular identifiers) observed for the gene in Set 1
  <li> set.2.umi -- the number of UMI (unique molecular identifiers) observed for the gene in Set 2
  <li> set.1.tpm -- gene expression in Set 1 in TPM (transcripts per million)
  <li> set.2.tpm -- gene expression in Set 2 in TPM (transcripts per million)
  <li> higher.expr -- which set has higher expression (TPM)
  <li> log2.ratio -- log2(TPM of higher expressing set / (TPM of lower expressing set + 1))
  <li> precision -- (# of cells expressing gene in higher expressing set) / (total # of cells expressing gene)
  <li> recall -- (# of cells expressing gene in higher expressing set) / (total # of cells in higher expressing set)
  <li> f.score -- geometric mean of precision and recall. genes are sorted by this metric
</ul>

{%highlight R%}
  ASEL.vs.ASER.DEG %>% head()
{%endhighlight%}

<table class = "table" style="overflow:auto;">
  <thead>
  <tr>
    <th>gene</th>
    <th>tank-1</th>
    <th>gcy-22</th>
    <th>gei-3</th>
    <th>gcy-3</th>
    <th>gcy-6</th>
    <th>T27C4.1</th>
  </tr>
</thead>
<tbody>
  <tr>
    <th>set.1.n.umi</th>
    <th>20</th>
    <th>0</th>
    <th>10</th>
    <th>0</th>
    <th>66</th>
    <th>30</th>
  </tr>
  <tr>
    <th>set.1.n.umi</th>
    <th>570</th>
    <th>129</th>
    <th>166</th>
    <th>147</th>
    <th>0</th>
    <th>164</th>
  </tr>
  <tr>
    <th>set.1.tpm</th>
    <th>1794.188</th>
    <th>0.000</th>
    <th>1095.928</th>
    <th>0.000</th>
    <th>6909.705</th>
    <th>3210.293</th>
  </tr>
  <tr>
    <th>set.2.tpm</th>
    <th>36169.778</th>
    <th>8862.360</th>
    <th>12076.301</th>
    <th>9593.506</th>
    <th>0.000</th>
    <th>11338.218</th>
  </tr>
  <tr>
    <th>higher.expr</th>
    <th>Set 2</th>
    <th>Set 2</th>
    <th>Set 2</th>
    <th>Set 2</th>
    <th>Set 1</th>
    <th>Set 2</th>
  </tr>
  <tr>
    <th>log2.ratio</th>
    <th>4.332578</th>
    <th>13.113475</th>
    <th>3.460638</th>
    <th>13.227842</th>
    <th>12.754408</th>
    <th>1.819968</th>
  </tr>
  <tr>
    <th>precision</th>
    <th>0.8536585</th>
    <th>1.0000000</th>
    <th>0.9090909</th>
    <th>1.0000000</th>
    <th>1.0000000</th>
    <th>0.6808511</th>
  </tr>
  <tr>
    <th>recall</th>
    <th>1.0000000</th>
    <th>0.8000000</th>
    <th>0.8571429</th>
    <th>0.7428571</th>
    <th>0.7297297	</th>
    <th>0.9142857</th>
  </tr>
  <tr>
    <th>f.score</th>
    <th>0.9210526</th>
    <th>0.8888889</th>
    <th>0.8823529</th>
    <th>0.8524590</th>
    <th>0.8437500</th>
    <th>0.7804878</th>
  </tr>
</tbody>
</table>

{% highlight R %}
ASEL.vs.ASER.DEG %>% filter(higher.expr == "Set 1")
  %>% head()
{% endhighlight %}


<table class = "table" style="overflow:auto;">
  <thead>
  <tr>
    <th>gene</th>
    <th>gcy-6</th>
    <th>gcy-17</th>
    <th>crh-1</th>
    <th>gcy-20</th>
    <th>gcy-7</th>
    <th>unc-44</th>
  </tr>
</thead>
<tbody>
  <tr>
    <th>set.1.n.umi</th>
    <th>66</th>
    <th>69</th>
    <th>58</th>
    <th>53</th>
    <th>39</th>
    <th>88</th>
  </tr>
  <tr>
    <th>set.2.n.umi</th>
    <th>0</th>
    <th>0</th>
    <th>12</th>
    <th>0</th>
    <th>0</th>
    <th>45</th>
  </tr>
  <tr>
    <th>set.1.n.tpm</th>
    <th>6909.705</th>
    <th>6553.874</th>
    <th>5675.336</th>
    <th>5339.037</th>
    <th>3709.660</th>
    <th>7647.438</th>
  </tr>
  <tr>
    <th>set.2.tpm</th>
    <th>0.000</th>
    <th>0.000</th>
    <th>800.504</th>
    <th>0.000</th>
    <th>0.000</th>
    <th>2915.088</th>
  </tr>
  <tr>
    <th>higher.expr</th>
    <th>Set 1</th>
    <th>Set 1</th>
    <th>Set 1</th>
    <th>Set 1</th>
    <th>Set 1</th>
    <th>Set 1</th>
  </tr>
  <tr>
    <th>log2.ratio</th>
    <th>12.754408</th>
    <th>12.678132</th>
    <th>2.823924</th>
    <th>12.382364</th>
    <th>11.857071</th>
    <th>1.390942</th>
  </tr>
  <tr>
    <th>precision</th>
    <th>1.0000000</th>
    <th>1.0000000</th>
    <th>0.8333333</th>
    <th>1.0000000</th>
    <th>1.0000000</th>
    <th>0.6000000</th>
  </tr>
  <tr>
    <th>recall</th>
    <th>0.7297297</th>
    <th>0.6216216</th>
    <th>0.6756757</th>
    <th>0.5945946</th>
    <th>0.5675676</th>
    <th>0.8108108</th>
  </tr>
  <tr>
    <th>f.score</th>
    <th>0.8437500</th>
    <th>0.7666667</th>
    <th>0.7462687</th>
    <th>0.7457627</th>
    <th>0.7241379</th>
    <th>0.6896552</th>
  </tr>
</tbody>
</table>

Running <em>two.set.differential.gene.test</em> with <em>formal = T</em> will compute q-value (p-value after Benjamini Hochberg correction). This is very slow however, so we recommend not using it for exploratory analysis and only using it after you've found something interesting and want to verify that the finding is statistically robust.

{% highlight R %}
system.time({
ASEL.vs.ASER.DEG = two.set.differential.gene.test(
    cds.neurons,
    is.neuron.type(cds.neurons, "ASEL"),
    is.neuron.type(cds.neurons, "ASER"),
    formal = T, cores = min(16, detectCores()))
})

{% endhighlight %}

<table style="overflow:auto;">
  <tr>
    <th>user</th>
    <th>system</th>
    <th>elapsed</th>
  </tr>
  <tr>
    <th>21.241</th>
    <th>3.599</th>
    <th>152.575</th>
  </tr>
</table>


{%highlight R%}
  ASEL.vs.ASER.DEG %>% head()
{%endhighlight%}

<table class = "table" style="overflow:auto;">
  <thead>
  <tr>
    <th>gene</th>
    <th>tank-1</th>
    <th>gcy-22</th>
    <th>gei-3</th>
    <th>gcy-3</th>
    <th>gcy-6</th>
    <th>T27C4.1</th>
  </tr>
</thead>
<tbody>
  <tr>
    <th>set.1.n.umi</th>
    <th>20</th>
    <th>0</th>
    <th>10</th>
    <th>0</th>
    <th>66</th>
    <th>30</th>
  </tr>
  <tr>
    <th>set.2.n.umi</th>
    <th>570</th>
    <th>129</th>
    <th>166</th>
    <th>147</th>
    <th>0</th>
    <th>164</th>
  </tr>
  <tr>
    <th>set.1.tpm</th>
    <th>1794.188</th>
    <th>0.000</th>
    <th>1095.928</th>
    <th>0.000</th>
    <th>6909.705</th>
    <th>3210.293</th>
  </tr>
  <tr>
    <th>set.2.tpm</th>
    <th>36169.778</th>
    <th>8862.360</th>
    <th>12076.301</th>
    <th>9593.506</th>
    <th>0.000</th>
    <th>11338.218</th>
  </tr>
  <tr>
    <th>higher.expr</th>
    <th>Set 2</th>
    <th>Set 2</th>
    <th>Set 2</th>
    <th>Set 2</th>
    <th>Set 1</th>
    <th>Set 2</th>
  </tr>
  <tr>
    <th>log2.ratio</th>
    <th>4.332578</th>
    <th>13.113475</th>
    <th>3.460638</th>
    <th>13.227842</th>
    <th>12.754408</th>
    <th>1.819968</th>
  </tr>
  <tr>
    <th>precision</th>
    <th>0.8536585</th>
    <th>1.0000000</th>
    <th>0.9090909</th>
    <th>1.0000000</th>
    <th>1.0000000</th>
    <th>0.6808511</th>
  </tr>
  <tr>
    <th>recall</th>
    <th>1.0000000</th>
    <th>0.8000000</th>
    <th>0.8571429</th>
    <th>0.7428571</th>
    <th>0.7297297</th>
    <th>0.9142857</th>
  </tr>
  <tr>
    <th>f.score</th>
    <th>0.9210526</th>
    <th>0.8888889</th>
    <th>0.8823529</th>
    <th>0.8524590</th>
    <th>0.8437500</th>
    <th>0.7804878</th>
  </tr>
</tbody>
</table>

{% highlight R %}
 touch.receptor.neurons.DEG  %>% filter(higher.expr == "Set 1") %>% head()
{% endhighlight %}

<table class = "table" style="overflow:auto;">
  <thead>
  <tr>
    <th>gene</th>
    <th>mec-17</th>
    <th>mec-18</th>
    <th>mtd-1</th>
    <th>mec-7</th>
    <th>mec-1</th>
    <th>mec-9</th>
  </tr>
</thead>
<tbody>
  <tr>
    <th>set.1.n.umi</th>
    <th>2720</th>
    <th>796</th>
    <th>443</th>
    <th>4418</th>
    <th>1717</th>
    <th>726</th>
  </tr>
  <tr>
    <th>set.2.n.umi</th>
    <th>118</th>
    <th>49</th>
    <th>15</th>
    <th>743</th>
    <th>695</th>
    <th>337</th>
  </tr>
  <tr>
    <th>set.1.tpm</th>
    <th>24033.028</th>
    <th>7040.671</th>
    <th>4025.528</th>
    <th>35936.308</th>
    <th>16304.188</th>
    <th>7088.606</th>
  </tr>
  <tr>
    <th>set.2.tpm</th>
    <th>39.202902</th>
    <th>17.963804</th>
    <th>5.513083</th>
    <th>239.162028</th>
    <th>328.822706</th>
    <th>182.571539</th>
  </tr>
  <tr>
    <th>higher.expr</th>
    <th>Set 1</th>
    <th>Set 1</th>
    <th>Set 1</th>
    <th>Set 1</th>
    <th>Set 1</th>
    <th>Set 1</th>
  </tr>
  <tr>
    <th>log2.ratio</th>
    <th>9.223503</th>
    <th>8.536321</th>
    <th>9.271622</th>
    <th>7.225290</th>
    <th>5.627408</th>
    <th>5.271088</th>
  </tr>
  <tr>
    <th>precision</th>
    <th>0.9009288</th>
    <th>0.9094828</th>
    <th>0.9476440</th>
    <th>0.5563771</th>
    <th>0.4609610</th>
    <th>0.5126263</th>
  </tr>
  <tr>
    <th>recall</th>
    <th>0.8712575</th>
    <th>0.6317365</th>
    <th>0.5419162</th>
    <th>0.9011976</th>
    <th>0.9191617</th>
    <th>0.6077844</th>
  </tr>
  <tr>
    <th>f.score</th>
    <th>0.8858447</th>
    <th>0.7455830</th>
    <th>0.6895238</th>
    <th>0.6880000</th>
    <th>0.6140000</th>
    <th>0.5561644</th>
  </tr>
</tbody>
</table>

<ul>
  <li> <em>tni-3</em> is a troponin that is expressed in body wall muscle (BWM) in the head, but not in the posterior.
  <li> <em>cwn-1</em> and <em>egl-20</em> are Wnt ligands that are expressed in posterior BWM, but not anterior BWM.
</ul>
  Using these genes as markers, let's look for differentially expressed genes between anterior and posterior BWM.


{% highlight R %}
BWM.anterior.vs.posterior.DEG =
  two.set.differential.gene.test(
    cds,
    is.cell.type(cds, "Body wall muscle") &
      expresses.gene(cds, "tni-3"),
    is.cell.type(cds, "Body wall muscle") &
      (expresses.gene(cds, "cwn-1") |
  expresses.gene(cds, "egl-20")))
{% endhighlight %}
<br>

{% highlight R %}
BWM.anterior.vs.posterior.DEG$higher.expr = ifelse(
    BWM.anterior.vs.posterior.DEG$higher.expr == "Set 1",
    "Anterior BWM", "Posterior BWM")
{% endhighlight %}

<br>
<div style = "overflow: scroll;">
<table class = "table">
  <thead>
  <tr>
    <th>gene</th>
    <th>set.1.n.umi</th>
    <th>set.2.n.umi</th>
    <th>set.1.tpm</th>
    <th>set.2.tpm</th>
    <th>higher.expr</th>
    <th>log2.ratio</th>
    <th>precision</th>
    <th>recall</th>
    <th>f.score</th>
  </tr>
</thead>
<tbody>
  <tr>
    <th>tni-13</th>
    <th>6981</th>
    <th>93</th>
    <th>3804.24742</th>
    <th>18.57887</th>
    <th>Anterior BWM</th>
    <th>7.602169</th>
    <th>0.9819890</th>
    <th>1.0000000</th>
    <th>0.9909127</th>
  </tr>
  <tr>
    <th>cwn-1</th>
    <th>59</th>
    <th>3113</th>
    <th>13.99762</th>
    <th>1311.17230</th>
    <th>Posterior BWM</th>
    <th>6.449980</th>
    <th>0.9821732</th>
    <th>0.8968992</th>
    <th>0.9376013</th>
  </tr>
  <tr>
    <th>T21B6.3</th>
    <th>13869</th>
    <th>271</th>
    <th>5966.03306</th>
    <th>60.20010</th>
    <th>Anterior BWM</th>
    <th>6.607094</th>
    <th>0.9553265</th>
    <th>0.8867624</th>
    <th>0.9197684</th>
  </tr>
  <tr>
    <th>glc-4</th>
    <th>684</th>
    <th>37</th>
    <th>310.53057</th>
    <th>11.4323046</th>
    <th>Anterior BWM	</th>
    <th>4.642570</th>
    <th>0.9506849</th>
    <th>0.2767145</th>
    <th>0.4286597</th>
  </tr>
  <tr>
    <th>tre-3	</th>
    <th>555</th>
    <th>5</th>
    <th>214.79033</th>
    <th>0.6369876</th>
    <th>Anterior BWM</th>
    <th>7.035742</th>
    <th>0.9852399</th>
    <th>0.2129187</th>
    <th>0.3501639</th>
  </tr>
  <tr>
    <th>F48E3.8</th>
    <th>402</th>
    <th>1</th>
    <th>144.96720</th>
    <th>0.1162907</th>
    <th>Anterior BWM</th>
    <th>7.020870</th>
    <th>0.9947917</th>
    <th>0.1523126</th>
    <th>0.2641770</th>
  </tr>
  <tr>
    <th>dpyd-1</th>
    <th>289</th>
    <th>6</th>
    <th>91.93291</th>
    <th>0.9050051</th>
    <th>Anterior BWM</th>
    <th>5.592715</th>
    <th>0.9740933</th>
    <th>0.1499203</th>
    <th>0.2598480</th>
  </tr>
  <tr>
    <th>ceh-34</th>
    <th>307</th>
    <th>2</th>
    <th>131.08432</th>
    <th>0.2528051</th>
    <th>Anterior BWM</th>
    <th>6.709189</th>
    <th>0.9885057</th>
    <th>0.1371611</th>
    <th>0.2408964</th>
  </tr>
  <tr>
    <th>seb-2</th>
    <th>201</th>
    <th>5</th>
    <th>85.08025	</th>
    <th>0.6848107</th>
    <th>Anterior BWM</th>
    <th>5.658166</th>
    <th>0.9750000</th>
    <th>0.1244019</th>
    <th>0.2206506</th>
  </tr>
  <tr>
    <th>sfrp-1</th>
    <th>335</th>
    <th>2</th>
    <th>145.18286</th>
    <th>0.2754131</th>
    <th>Anterior BWM</th>
    <th>6.830763</th>
    <th>0.9863946</th>
    <th>0.1156300</th>
    <th>0.2069950</th>
  </tr>
  <tr>
    <th>F35C11.5</th>
    <th>168</th>
    <th>1</th>
    <th>67.97453</th>
    <th>0.5230727</th>
    <th>Anterior BWM</th>
    <th>5.479937</th>
    <th>0.9923664</th>
    <th>0.1036683</th>
    <th>0.1877256</th>
  </tr>
</tbody>
</table>
</div>

<em>R102.2</em> is expressed in  <a href = "http://www.wormbase.org/species/c_elegans/gene/WBGene00011289"> nine specific pairs of ciliated sensory neurons </a>.
We have identified some of these in this dataset, but others remain ambiguous.

{% highlight R %}
  show.expr.info("R102.2", "neuron type") %>% head(8)
{% endhighlight %}

<table class = "table" style="overflow:auto;">
  <thead>
  <tr>
    <th>facet</th>
    <th>tpm</th>
    <th>prop.cells.expr</th>
    <th>n.umi</th>
    <th>total.n.umi.for.facet</th>
  </tr>
</thead>
<tbody>
  <tr>
    <th>Cluster 21</th>
    <th>9807.74014</th>
    <th>0.91588785</th>
    <th>446</th>
    <th>49587</th>
  </tr>
  <tr>
    <th>Cluster 16</th>
    <th>8266.64832</th>
    <th>0.66025641</th>
    <th>363</th>
    <th>47719</th>
  </tr>
  <tr>
    <th>ASK</th>
    <th>6720.61429</th>
    <th>0.81944444</th>
    <th>155</th>
    <th>24157</th>
  </tr>
  <tr>
    <th>ASI/ASJ</th>
    <th>6482.22237</th>
    <th>0.74358974</th>
    <th>239</th>
    <th>40396</th>
  </tr>
  <tr>
    <th>ASG</th>
    <th>2121.04289</th>
    <th>0.39534884</th>
    <th>48</th>
    <th>26295</th>
  </tr>
  <tr>
    <th>ASEL</th>
    <th>1670.50501</th>
    <th>0.37837838</th>
    <th>17</th>
    <th>11042</th>
  </tr>
  <tr>
    <th>ASER</th>
    <th>795.91129</th>
    <th>0.28571429</th>
    <th>16</th>
    <th>15324</th>
  </tr>
  <tr>
    <th>AWB/AWC</th>
    <th>85.09161</th>
    <th>0.02380952</th>
    <th>4</th>
    <th>23186</th>
  </tr>
</tbody>
</table>

{% highlight R %}
plot.expr(cds.neurons, "R102.2")
{% endhighlight %}

<div class= "text-center">
  <img src= "{{site.baseurl}}/images/Cao_2017_3.png" width = 600>
</div>

<em> two.set.differential.gene.test </em> can be used to find DEG between the mystery R102.2(+) clusters and other ciliated sensory neurons. If you can figure out what these clusters correspond to, let us know!

{% highlight R %}
neuron.cluster.21.vs.other.CSN.DEG =
  two.set.differential.gene.test(
    cds.neurons,
    is.neuron.type(cds.neurons, "Cluster 21"),
    !is.neuron.type(cds.neurons, "Cluster 21") &
    is.cell.type(cds.neurons, "Ciliated sensory neurons"))

neuron.cluster.21.vs.other.CSN.DEG$higher.expr = ifelse(
  neuron.cluster.21.vs.other.CSN.DEG$higher.expr == "Set 1",
  "Cluster 21", "Other CSN")
{% endhighlight %}

<br>

 {% highlight R %}
  neuron.cluster.21.vs.other.CSN.DEG %>%
    filter(higher.expr == "Cluster 21",
  precision >= 0.8) %>% head(10)
 {% endhighlight %}

<div style = "overflow:scroll">
 <table class = "table">
   <thead>
   <tr>
     <th>gene</th>
     <th>set.1.n.umi</th>
     <th>set.2.n.umi</th>
     <th>set.1.tpm</th>
     <th>set.2.tpm</th>
     <th>higher.expr</th>
     <th>log2.ratio</th>
     <th>precision</th>
     <th>recall</th>
     <th>f.score</th>
   </tr>
 </thead>
 <tbody>
   <tr>
     <th>C39D10.2</th>
     <th>226</th>
     <th>1</th>
     <th>5237.910</th>
     <th>1.667334</th>
     <th>Cluster 21</th>
     <th>10.939377</th>
     <th>0.9880952</th>
     <th>0.7757009</th>
     <th>0.8691099</th>
   </tr>
   <tr>
     <th>T09B9.3</th>
     <th>197</th>
     <th>0</th>
     <th>3836.121</th>
     <th>0.000000</th>
     <th>Cluster 21</th>
     <th>11.905433</th>
     <th>1.0000000</th>
     <th>0.6074766</th>
     <th>0.7558140</th>
   </tr>
   <tr>
     <th>F15A4.5</th>
     <th>157</th>
     <th>15</th>
     <th>3677.812</th>
     <th>67.963116</th>
     <th>Cluster 21</th>
     <th>5.736879</th>
     <th>0.8923077</th>
     <th>0.5420561</th>
     <th>0.6744186</th>
   </tr>
   <tr>
     <th>flp-25</th>
     <th>100</th>
     <th>7</th>
     <th>1775.299</th>
     <th>44.020086</th>
     <th>Cluster 21</th>
     <th>5.301349</th>
     <th>0.9137931</th>
     <th>0.4953271</th>
     <th>0.6424242</th>
   </tr>
   <tr>
     <th>C18H7.6</th>
     <th>118</th>
     <th>0</th>
     <th>2385.147</th>
     <th>0.000000</th>
     <th>Cluster 21</th>
     <th>11.219863</th>
     <th>1.0000000</th>
     <th>0.4579439</th>
     <th>0.6282051</th>
   </tr>
   <tr>
     <th>cdh-3</th>
     <th>79</th>
     <th>11</th>
     <th>1811.618</th>
     <th>69.489017</th>
     <th>Cluster 21</th>
     <th>4.683736</th>
     <th>0.8809524</th>
     <th>0.3457944</th>
     <th>0.4966443</th>
   </tr>
   <tr>
     <th>K04D7.6</th>
     <th>65</th>
     <th>0</th>
     <th>1110.520</th>
     <th>0.000000</th>
     <th>Cluster 21</th>
     <th>10.117019</th>
     <th>1.0000000</th>
     <th>0.2710280</th>
     <th>0.4264706</th>
   </tr>
   <tr>
     <th>C29F4.3</th>
     <th>53</th>
     <th>0</th>
     <th>1067.368</th>
     <th>0.000000</th>
     <th>Cluster 21</th>
     <th>10.059842</th>
     <th>1.0000000</th>
     <th>0.2523364</th>
     <th>0.4029851</th>
   </tr>
   <tr>
     <th>K02E2.1</th>
     <th>39</th>
     <th>0</th>
     <th>1084.714</th>
     <th>0.000000</th>
     <th>Cluster 21</th>
     <th>10.083099</th>
     <th>1.0000000</th>
     <th>0.2523364</th>
     <th>0.4029851</th>
   </tr>
   <tr>
     <th>dhs-9</th>
     <th>67</th>
     <th>15</th>
     <th>987.323</th>
     <th>53.024505</th>
     <th>Cluster 21</th>
     <th>4.191836</th>
     <th>0.8484848	</th>
     <th>0.2616822</th>
     <th>0.4000000</th>
   </tr>
 </tbody>
 </table>
 </div>

{% highlight R %}
neuron.cluster.16.vs.other.CSN.DEG =
  two.set.differential.gene.test(
    cds.neurons,
    is.neuron.type(cds.neurons, "Cluster 16"),
    !is.neuron.type(cds.neurons, "Cluster 16") &
        is.cell.type(cds.neurons,
      "Ciliated sensory neurons"))

neuron.cluster.16.vs.other.CSN.DEG$higher.expr = ifelse(
  neuron.cluster.16.vs.other.CSN.DEG$higher.expr == "Set 1",
  "Cluster 16", "Other CSN")
{% endhighlight %}

<br>

{% highlight R %}

  neuron.cluster.16.vs.other.CSN.DEG %>%
      filter(higher.expr == "Cluster 16") %>% head(10)

{% endhighlight %}
<div style = "overflow:scroll">
<table class = "table" style="overflow:auto;">
  <thead>
  <tr>
    <th>gene</th>
    <th>set.1.n.umi</th>
    <th>set.2.n.umi</th>
    <th>set.1.tpm</th>
    <th>set.2.tpm</th>
    <th>higher.expr</th>
    <th>log2.ratio</th>
    <th>precision</th>
    <th>recall</th>
    <th>f.score</th>
  </tr>
</thead>
<tbody>
  <tr>
    <th>F27C1.11</th>
    <th>223</th>
    <th>367</th>
    <th>5302.959</th>
    <th>1576.21303</th>
    <th>Cluster 16</th>
    <th>1.7494202</th>
    <th>0.3537118</th>
    <th>0.5192308</th>
    <th>0.4207792</th>
  </tr>
  <tr>
    <th>W05F2.7</th>
    <th>137</th>
    <th>309</th>
    <th>3007.794</th>
    <th>1054.38784</th>
    <th>Cluster 16</th>
    <th>1.5109326</th>
    <th>0.3564356</th>
    <th>0.4615385</th>
    <th>0.4022346</th>
  </tr>
  <tr>
    <th>M04B2.6</th>
    <th>117</th>
    <th>4</th>
    <th>2144.256</th>
    <th>14.78412</th>
    <th>Cluster 16</th>
    <th>7.0858594</th>
    <th>0.9285714</th>
    <th>0.2500000</th>
    <th>0.3939394</th>
  </tr>
  <tr>
    <th>ocr-2</th>
    <th>100</th>
    <th>81</th>
    <th>2408.536</th>
    <th>294.50719</th>
    <th>Cluster 16</th>
    <th>3.0268916</th>
    <th>0.5465116</th>
    <th>0.3012821</th>
    <th>0.3884298</th>
  </tr>
  <tr>
    <th>osm-10</th>
    <th>66</th>
    <th>32</th>
    <th>1209.716</th>
    <th>124.87432</th>
    <th>Cluster 16</th>
    <th>3.2646129</th>
    <th>0.7222222</th>
    <th>0.2500000</th>
    <th>0.3714286</th>
  </tr>
  <tr>
    <th>R102.2</th>
    <th>363</th>
    <th>926</th>
    <th>8266.648</th>
    <th>3753.48748</th>
    <th>Cluster 16</th>
    <th>1.1386865</th>
    <th>0.2524510</th>
    <th>0.6602564</th>
    <th>0.3652482</th>
  </tr>
  <tr>
    <th>T01D3.1</th>
    <th>87</th>
    <th>122</th>
    <th>1849.955</th>
    <th>382.64946</th>
    <th>Cluster 16</th>
    <th>2.2696298</th>
    <th>0.4112903</th>
    <th>0.3269231</th>
    <th>0.3642857</th>
  </tr>
  <tr>
    <th>ida-1</th>
    <th>337</th>
    <th>1211</th>
    <th>7843.130</th>
    <th>5636.27821</th>
    <th>Cluster 16</th>
    <th>0.4764307</th>
    <th>0.2234848</th>
    <th>0.7564103</th>
    <th>0.3450292</th>
  </tr>
  <tr>
    <th>lap-2</th>
    <th>132</th>
    <th>55</th>
    <th>2344.214</th>
    <th>231.93208</th>
    <th>Cluster 16</th>
    <th>3.3311228</th>
    <th>0.5263158</th>
    <th>0.2564103</th>
    <th>0.3448276</th>
  </tr>
  <tr>
    <th>rps-11</th>
    <th>83</th>
    <th>230</th>
    <th>2013.967</th>
    <th>1036.83972</th>
    <th>Cluster 16</th>
    <th>0.9564565</th>
    <th>0.2863636	</th>
    <th>0.4038462</th>
    <th>0.3351064</th>
  </tr>
</tbody>
</table>
</div>


<h2>Use Case 4: Re-clustering cell subsets using t-SNE</h2>

<p>In the Cao et al. paper, we found that several of the neuron t-SNE clusters expressed markers of cholinergic neurons such as unc-17, cho-1, and cha-1.
Let's perform a sub-clustering of these cholinergic neurons to see if we can get better separation.
First, we'll identify which clusters are enriched for cells that express cholinergic markers.</p>

<p>
{% highlight R %}
pData(cds.neurons)$any.cholinergic.marker =
    (expresses.gene(cds.neurons, "unc-17") +
     expresses.gene(cds.neurons, "cho-1") +
     expresses.gene(cds.neurons, "cha-1")) > 0
{%endhighlight%}
</p>

<p>
{% highlight R %}
pData(cds.neurons) %>%
    group_by(Cluster) %>%
    summarize(
        n.total = n(), n.cholinergic =
          sum(any.cholinergic.marker),
        prop.cholinergic = n.cholinergic / n.total) %>%
    inner_join(
        unique(pData(cds.neurons)[, c("Cluster",
            "neuron.type")]),
        by = "Cluster") %>%
    arrange(-prop.cholinergic) %>%
    head(15)
{%endhighlight%}
</p>


<table class = "table" style="overflow:auto;">
  <thead>
  <tr>
    <th>Cluster</th>
    <th>n.total</th>
    <th>n.cholinergic</th>
    <th>prop.cholinergic</th>
    <th>neuron.type</th>
  </tr>
</thead>
<tbody>
  <tr>
    <th>29</th>
    <th>305</th>
    <th>155</th>
    <th>0.5081967</th>
    <th>Cholinergic (29)</th>
  </tr>
  <tr>
    <th>23</th>
    <th>45</th>
    <th>19</th>
    <th>0.4222222</th>
    <th>Cholinergic (23)</th>
  </tr>
  <tr>
    <th>3</th>
    <th>385</th>
    <th>131</th>
    <th>0.3402597</th>
    <th>Cholinergic (3)</th>
  </tr>
  <tr>
    <th>26</th>
    <th>261</th>
    <th>81</th>
    <th>0.3103448</th>
    <th>Cholinergic (26)</th>
  </tr>
  <tr>
    <th>35</th>
    <th>58</th>
    <th>16</th>
    <th>0.2758621</th>
    <th>Cholinergic (35)</th>
  </tr>
  <tr>
    <th>36</th>
    <th>128</th>
    <th>35</th>
    <th>0.2734375</th>
    <th>Cholinergic (36)</th>
  </tr>
  <tr>
    <th>15</th>
    <th>65</th>
    <th>16</th>
    <th>0.2461538</th>
    <th>Cholinergic (15)</th>
  </tr>
  <tr>
    <th>12</th>
    <th>68</th>
    <th>14</th>
    <th>0.2058824</th>
    <th>DVA</th>
  </tr>
  <tr>
    <th>24</th>
    <th>188</th>
    <th>38</th>
    <th>0.2021277</th>
    <th>Cholinergic (24)</th>
  </tr>
  <tr>
    <th>8</th>
    <th>117</th>
    <th>23</th>
    <th>0.1965812</th>
    <th>ASI/ASJ</th>
  </tr>
  <tr>
    <th>11</th>
    <th>1998</th>
    <th>387</th>
    <th>0.1936937</th>
    <th>Cholinergic (11)</th>
  </tr>
  <tr>
    <th>6</th>
    <th>160</th>
    <th>21</th>
    <th>0.1312500</th>
    <th>SDQ/ALN/PLN</th>
  </tr>
  <tr>
    <th>25</th>
    <th>363</th>
    <th>43</th>
    <th>0.1184573</th>
    <th>Cluster 25</th>
  </tr>
  <tr>
    <th>41</th>
    <th>211</th>
    <th>24</th>
    <th>0.1137441</th>
    <th>Doublets</th>
  </tr>
  <tr>
    <th>16</th>
    <th>156</th>
    <th>17</th>
    <th>0.1089744</th>
    <th>Cluster 16</th>
  </tr>
</tbody>
</table>
<p>

{%highlight R%}
cholinergic.clusters = c(29, 23, 3, 26, 35, 36, 15, 24, 11)
plot_cell_clusters(cds.neurons, color =
  "Cluster %in% cholinergic.clusters",
  cell_size = 0.2)
{% endhighlight %}

<div class= "text-center">
  <img src= "{{site.baseurl}}/images/Cao_2017_4.png" width = 600>
</div>

{% highlight R %}
cds.cholinergic =
    cds.neurons[, pData(cds.neurons)$Cluster %in%
        cholinergic.clusters]
cat(ncol(cds.cholinergic), "cells in the cds subset", "\n")

cds.cholinergic = estimateSizeFactors(cds.cholinergic)
cds.cholinergic = estimateDispersions(cds.cholinergic)
    # above line would take a lot of memory for larger cell sets
cds.cholinergic = detectGenes(cds.cholinergic, 0.1)
{% endhighlight %}

<p> 3433 cells in the cds subset </p>

<p> The next step is to run a new t-SNE dimensionality reduction on this subset of cells. </p>

<p>
  {% highlight R %}
system.time({
cds.cholinergic = reduceDimension(
    cds.cholinergic, max_components = 2, norm_method = "log",
    num_dim = 20, reduction_method = 'tSNE', verbose = T)
})

pData(cds.cholinergic)$tsne_1 = reducedDimA(cds.cholinergic)[1,]
pData(cds.cholinergic)$tsne_2 = reducedDimA(cds.cholinergic)[2,]
  {% endhighlight %}
</p>

<table style="overflow:auto;">
  <tr>
    <th> user </th>
    <th> system </th>
    <th> elapsed </th>
  </tr>
  <tr>
    <th> 164.398 </th>
    <th> 2.898 </th>
    <th> 168.483 </th>
  </tr>
</table>

<p> 20 principal components looks like it's enough for this data.
  If you don't see an elbow in the scree plot, that means you've used too few principal components. </p>

{% highlight R %}
  plot_pc_variance_explained(cds.cholinergic)
{% endhighlight %}

<div class= "text-center">
  <img src= "{{site.baseurl}}/images/Cao_2017_5.png" width = 600>
</div>

{% highlight R %}
ggplot(pData(cds.cholinergic), aes(x = tsne_1, y = tsne_2)) +
    geom_point(size = 0.1) +
    monocle:::monocle_theme_opts()
{% endhighlight %}

<div class= "text-center">
  <img src= "{{site.baseurl}}/images/Cao_2017_6.png" width = 600>
</div>

<p> Now we'll cluster the cells in the t-SNE space using density peak clustering.
In density peak clustering, rho measures to the local density of cells and delta measures the minimum distance to a region of higher local density.
Go to <a href = "http://science.sciencemag.org/content/344/6191/1492"> this site </a> for more details.
You want to set thresholds on rho and delta that enclose the outlier points in the scatter plot.
In my experience, it's usually better to over-cluster than to under-cluster. </p>

{% highlight R %}
  cds.cholinergic = clusterCells_Density_Peak(cds.cholinergic)
{% endhighlight %}

<p> Distance cutoff calculated to 2.694535 </p>

{% highlight R %}
  plot_rho_delta(cds.cholinergic,
    rho_threshold = 10,
    delta_threshold = 6)
{% endhighlight %}

<div class= "text-center">
  <img src= "{{site.baseurl}}/images/Cao_2017_7.png" width = 600>
</div>

{% highlight R %}
cds.cholinergic = clusterCells_Density_Peak(cds.cholinergic,
    rho_threshold = 10, delta_threshold = 6, skip_rho_sigma = T)
plot_cell_clusters(cds.cholinergic, cell_size = 0.2)
{% endhighlight %}

<div class= "text-center">
  <img src= "{{site.baseurl}}/images/Cao_2017_8.png" width = 600>
</div>

<p> Looks like there are many distinct clusters.
  Recall that we input only 9 clusters from the cds.neurons t-SNE into this re-clustering.
  The re-clustering appears to have revealed new potential distinct cell types. </p>

<p> Now would be a good time to save your progress </p>

{% highlight R %}
  save.image("my_analysis_Cao_et_al_data.RData")
{% endhighlight %}

<p> Let's find marker genes for each cluster. </p>

{% highlight R %}
  cholinergic.clusters =
      sort(as.integer(unique(pData(cds.cholinergic)$Cluster)))
  cholinergic.clusters
{% endhighlight %}

<p> 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 </p>

{% highlight R %}
DEG.results = lapply(cholinergic.clusters,
    function(this.cluster) {
  message("Finding markers for cluster ", this.cluster)
    cbind(
        two.set.differential.gene.test(
            cds.cholinergic,
            pData(cds.cholinergic)$Cluster == this.cluster,
            pData(cds.cholinergic)$Cluster != this.cluster),
        data.frame(cluster = this.cluster))
})
{% endhighlight %}

{% highlight R %}
cholinergic.cluster.markers =
    do.call(rbind, DEG.results) %>%
        filter(higher.expr == "Set 1") %>%
  select(
      cluster, gene,
      cluster.n.umi = set.1.n.umi,
      other.n.umi = set.2.n.umi,
      cluster.tpm = set.1.tpm,
      other.tpm = set.2.tpm,
      log2.ratio, precision, recall, f.score) %>%
  arrange(-f.score)
{% endhighlight %}

{% highlight R %}
  cholinergic.cluster.markers %>%
      filter(precision >= 0.5) %>% head(10)
{% endhighlight %}

<div style = "overflow:scroll">
<table class = "table" >
  <thead>
  <tr>
    <th>Cluster</th>
    <th>gene</th>
    <th>cluster.n.umi</th>
    <th>cluster.n.omi</th>
    <th>cluster.tpm</th>
    <th>other.tpm</th>
    <th>log2.ratio</th>
    <th>precision</th>
    <th>recall</th>
    <th>f.score</th>
  </tr>
</thead>
<tbody>
  <tr>
    <th>7</th>
    <th>B0432.14</th>
    <th>194</th>
    <th>11</th>
    <th>10886.411</th>
    <th>5.8332893</th>
    <th>10.637661</th>
    <th>0.9047619</th>
    <th>0.9500000</th>
    <th>0.9268293</th>
  </tr>
  <tr>
    <th>13</th>
    <th>nlp-42</th>
    <th>299</th>
    <th>15</th>
    <th>22725.799</th>
    <th>17.1886827</th>
    <th>10.287074</th>
    <th>0.8833333</th>
    <th>0.9137931</th>
    <th>0.8983051</th>
  </tr>
  <tr>
    <th>13</th>
    <th>nlp-42</th>
    <th>299</th>
    <th>15</th>
    <th>22725.799</th>
    <th>17.1886827</th>
    <th>10.287074</th>
    <th>0.8833333</th>
    <th>0.9137931</th>
    <th>0.8983051</th>
  </tr>
  <tr>
    <th>21</th>
    <th>flp-12</th>
    <th>1379</th>
    <th>297</th>
    <th>17963.487</th>
    <th>366.8397815</th>
    <th>5.609846</th>
    <th>0.7282609</th>
    <th>0.8777293</th>
    <th>0.7960396</th>
  </tr>
  <tr>
    <th>13</th>
    <th>T04C12.3</th>
    <th>173</th>
    <th>25</th>
    <th>12067.242</th>
    <th>22.8874212</th>
    <th>8.980629</th>
    <th>0.7407407</th>
    <th>0.6896552</th>
    <th>0.7142857</th>
  </tr>
  <tr>
    <th>22</th>
    <th>lgc-39	</th>
    <th>289</th>
    <th>93</th>
    <th>5300.848</th>
    <th>120.7416859</th>
    <th>5.444328</th>
    <th>0.6346154</th>
    <th>0.6839378</th>
    <th>0.6583541</th>
  </tr>
  <tr>
    <th>1</th>
    <th>sem-2</th>
    <th>128</th>
    <th>63</th>
    <th>6240.354</th>
    <th>52.4115889</th>
    <th>6.868331</th>
    <th>0.6615385</th>
    <th>0.6056338</th>
    <th>0.6323529</th>
  </tr>
  <tr>
    <th>2</th>
    <th>vglu-2</th>
    <th>30</th>
    <th>1</th>
    <th>2356.216</th>
    <th>0.6977763</th>
    <th>10.438609</th>
    <th>0.9523810</th>
    <th>0.4444444</th>
    <th>0.6060606</th>
  </tr>
  <tr>
    <th>15</th>
    <th>Y48C3A.5</th>
    <th>339</th>
    <th>145</th>
    <th>6989.459</th>
    <th>148.0684921</th>
    <th>5.551134</th>
    <th>0.5747664</th>
    <th>0.6243655</th>
    <th>0.5985401</th>
  </tr>
  <tr>
    <th>10</th>
    <th>glb-17</th>
    <th>81</th>
    <th>19</th>
    <th>4790.632</th>
    <th>16.9955396</th>
    <th>8.056433</th>
    <th>0.7352941</th>
    <th>0.5000000</th>
    <th>0.5952381</th>
  </tr>
  <tr>
    <th>9</th>
    <th>nlp-5</th>
    <th>34</th>
    <th>24</th>
    <th>3552.050</th>
    <th>21.1320088</th>
    <th>7.326374</th>
    <th>0.5500000</th>
    <th>0.6470588</th>
    <th>0.5945946</th>
  </tr>
</tbody>
</table>
</div>
<p>


<p>In the previous examples for using <em>plot.expr</em>, the function used pre-computed expression tables for <em>cds</em> and <em>cds.neurons</em> to show which cell types / neuron types were expressing a given gene.
  The following code will set up <em>plot.expr</em> so it can show which density peak clusters from this re-clustering analysis express a given gene.</p>

  {% highlight R %}
cholinergic.expr.info = get.expr.info.by.facet(cds.cholinergic,
  "Cluster")
  {% endhighlight %}

  {% highlight R %}
plot.expr(cds.cholinergic, "lgc-39", expr.info =
  cholinergic.expr.info)
  {% endhighlight %}

  <div class= "text-center">
    <img src= "{{site.baseurl}}/images/Cao_2017_9.png" width = 600>
  </div>

  {% highlight R %}
    plot.expr(cds.cholinergic, "vglu-2", expr.info =
      cholinergic.expr.info)
  {% endhighlight %}

  <div class= "text-center">
    <img src= "{{site.baseurl}}/images/Cao_2017_10.png" width = 600>
  </div>

  {% highlight R %}
    plot.expr(cds.cholinergic, "unc-17", expr.info =
      cholinergic.expr.info)
  {% endhighlight %}

  <div class= "text-center">
    <img src= "{{site.baseurl}}/images/Cao_2017_11.png" width = 600>
  </div>

</div>
</div>
</div>

<br>
<br>
<br>
<br>
<br>
<br>

{% include footer.html %}

</body>

</html>
